### ISSUE TITLE
Print function for use within loops

### ISSUE THREAD

**Issue by casperdcl** [tqdm/tqdm#116]
as discussed in #113

**Comment by lrq3000** [comment]
Specification: the tqdm.print method should manage the printing of messages
in the console without messing with tqdm bars, even if nested or positioned
bars are used.
My idea is to just clear all tqdm bars, place cursor to pos 0, print the
messages, and the new position will be the new position 0, so we just have
to force refresh the display of all bars.
The only thing we have to be wary of is to make sure all bars are cleaned
and moved. That's why i propose to implement the print as a class method,
but if there's another way as elegant I'm ok.
Le 21 Jan. 2016 22:31, "Casper da Costa-Luis" <notifications@github.com> a
écrit :

as discussed in #113 <https://github.com/tqdm/tqdm/issues/113>

—
Reply to this email directly or view it on GitHub
<https://github.com/tqdm/tqdm/issues/116>.

**Comment by lrq3000** [comment]
I tried my idea but I've run into an issue: it doesn't work with tqdm.__iter__() because I have no way to force the refresh printing of iterable-based bars.

To detail my idea, I wanted to add something like this:

```
class tqdm():

    def refresh(self):
        self.sp(self.format(self.n, self.total, ...))

    @classmethod
    def print(cls, message, file=sys.stdout):
        # clear the bars
        for instance in reverse(cls._instances):
            instance.sp()
        # print message at current cursor position
        file.write(message)
        file.write("\n")
        # redraw all bars afterwards
        for instance in cls._instances:
            instance.refresh()
```

This should work very well for manually updated tqdm, but not for iterable, because we cannot refresh this way (self.n is not stored).

So either we have to remove this optimization (maybe this won't impact performance?), or we have to find a way to put some kind of remotely triggerable interrupt in the __iter__() method so that we can remotely force a refresh.

**Comment by lrq3000** [comment]
Found a way to elegantly (I think) implement this feature. See #120 for the rest of the discussion.

**Comment by lrq3000** [comment]
Merged in #150.
